pkg/actuators/machine/machine.go

```diff
@@ -54,80 +53,178 @@ func newMachineScope(
+			}
+			klog.Infof("Waiting for disk to become OK...")
+			updatedDisk, err = updatedDisk.WaitForOK()
+			if err != nil {
+				return errors.Wrapf(err, "Failed to wait for disk %s to return to OK status. (%v)", disk.ID())
+			}
+		}
+	}
+
+	// handleNics reattachment
+	if ms.machineProviderSpec.NetworkInterfaces != nil && len(ms.machineProviderSpec.NetworkInterfaces) > 0 {
+		nics, err := instance.ListNICs()
+		if err != nil {
+			errors.Wrapf(err, "failed to list NICs on VM %s", instance.ID())
+		}
+
+		//remove all the nics from the VM instance
+		for _, nic := range nics {
+			if err := nic.Remove(); err != nil {
+				errors.Wrapf(err, "failed to remove NIC %s", nic.ID())
+			}
+		}
+
+		//re-create NICs According to the machinespec
+		for i, nic := range ms.machineProviderSpec.NetworkInterfaces {
+			_, err := instance.CreateNIC(fmt.Sprintf("nic%d", i+1), nic.VNICProfileID, ovirtC.CreateNICParams())
+
+			if err != nil {
+				return errors.Wrap(err, "failed to create network interface")
+			}
+		}
+	}
+
+	if isAutoPinning {
+		if ms.machineProviderSpec.AutoPinningPolicy == "resize_and_pin" {
+			err = ms.ovirtClient.AutoOptimizeVMCPUPinningSettings(instance.ID(), true)
+			if err != nil {
+				errors.Wrapf(err, "failed to Optimize CPU pinning settings for VM %s", instance.ID())
+			}
+		}
+	}
+
+	// TODO: attach Tag to a VM using tagName instead of ID - https://github.com/oVirt/go-ovirt-client/issues/135
+	// TODO: handleAffinityGroups -
+
 	// Start the VM
-	err = ms.ovirtClient.StartVM(instance.MustId())
+	err = ms.ovirtClient.StartVM(instance.ID())
 	if err != nil {
 		return errors.Wrap(err, "error running oVirt VM")
 	}
 
 	// Wait till running
-	err = ms.waitTillVMReachStatus(ovirtsdk.VMSTATUS_UP)
+	_, err = ms.ovirtClient.WaitForVMStatus(instance.ID(), ovirtC.VMStatusUp)
 	if err != nil {
 		return errors.Wrap(err, "Error waiting for oVirt VM to be UP")
 	}
 	return nil
 }
 
-func (ms *machineScope) waitTillVMReachStatus(status ovirtsdk.VmStatus) error {
-	return util.PollImmediate(retryIntervalInstanceStatus, timeoutInstanceCreate, func() (bool, error) {
-		instance, err := ms.ovirtClient.GetVMByMachine(*ms.machine)
-		if err != nil {
-			return false, err
-		}
-		return instance.MustStatus() == status, nil
-	})
-}
-
 // exists returns true if machine exists.
 func (ms *machineScope) exists() (bool, error) {
-	instance, err := ms.ovirtClient.GetVMByMachine(*ms.machine)
+	vm, err := ms.ovirtClient.GetVMByName(ms.machine.Name)
 	if err != nil {
 		return false, errors.Wrap(err, "error finding VM by name")
 	}
-	return instance != nil, nil
+	return vm != nil, nil
 }
 
 // delete deletes the VM which corresponds with the machine object from the oVirt engine
 func (ms *machineScope) delete() error {
-	instance, err := ms.ovirtClient.GetVMByMachine(*ms.machine)
+	vm, err := ms.ovirtClient.GetVMByName(ms.machine.Name)
 	if err != nil {
 		return errors.Wrap(err, "error finding VM by name")
 	}
-	if instance == nil {
+
+	if vm == nil {
```

Comment:
This should never be the case, GetVMByName should always return a VM or an error. You need to check for the proper error code here.

Reply:
removed this check